cmd/ocp-build-data-enforcer/main.go

```diff
@@ -0,0 +1,304 @@
+	}
+
+	errGroup := &errgroup.Group{}
+	for idx := range configs {
+		idx := idx
+		errGroup.Go(func() error {
+			processDockerfile(configs[idx])
+			return nil
+		})
+	}
+	if err := errGroup.Wait(); err != nil {
+		logrus.WithError(err).Fatal("Processing failed")
+	}
+
+	if err := utilerrors.NewAggregate(errs); err != nil {
+		for _, err := range err.Errors() {
+			logrus.WithError(err).Error("Encountered error")
+		}
+		logrus.Fatal("Encountered errors")
+	}
+	logrus.Infof("Processed %d configs", len(configs))
+}
+
+// TODO (alvaroaleman): This is incomplete
+func dereferenceConfig(config *ocpImageConfig, streamMap streamMap, groupYAML groupYAML) error {
+	if replacement, hasReplacement := streamMap[config.From.Stream]; hasReplacement {
+		config.From.Stream = replacement.UpstreamImage
+	}
+	for blder := range config.From.Builder {
+		if replacement, hasReplacement := streamMap[config.From.Builder[blder].Stream]; hasReplacement {
+			config.From.Builder[blder].Stream = replacement.UpstreamImage
+		}
+	}
+	if config.Content.Source.Alias != "" {
+		if _, hasReplacement := groupYAML.Sources[config.Content.Source.Alias]; !hasReplacement {
+			return fmt.Errorf("groups.yaml has no replacement for alias %s", config.Content.Source.Alias)
+		}
+		config.Content.Source.Git = &ocpImageConfigSourceGit{}
+		*config.Content.Source.Git = groupYAML.Sources[config.Content.Source.Alias]
+	}
+
+	return nil
+}
+
+func processDockerfile(config ocpImageConfig) {
+	orgRepo := config.orgRepo()
+	log := logrus.WithField("file", config.SourceFileName).WithField("org/repo", orgRepo)
+	split := strings.Split(orgRepo, "/")
+	if n := len(split); n != 2 {
+		log.Errorf("splitting orgRepo didn't yield 2 but %d results", n)
+		return
+	}
+	if split[0] == "openshift-priv" {
+		log.Trace("Ignoring repo in openshift-priv org")
+		return
+	}
+	org, repo := split[0], split[1]
+	getter := github.FileGetterFactory(org, repo, "release-4.6")
+
+	log = log.WithField("dockerfile", config.dockerfile())
+	data, err := getter(config.dockerfile())
+	if err != nil {
+		log.WithError(err).Error("Failed to get dockerfile")
+		return
+	}
+	if len(data) == 0 {
+		log.Error("dockerfile is empty")
+		return
+	}
+
+	updated, hasDiff, err := updateDockerfile(data, config)
+	if err != nil {
+		log.WithError(err).Error("Failed to update Dockerfile")
+		return
+	}
+	if !hasDiff {
+		return
+	}
+	diff := difflib.UnifiedDiff{
+		A:        difflib.SplitLines(string(data)),
+		B:        difflib.SplitLines(string(updated)),
+		FromFile: "original",
+		ToFile:   "updated",
+		Context:  3,
+	}
+	diffStr, err := difflib.GetUnifiedDiffString(diff)
+	if err != nil {
+		log.WithError(err).Error("Failed to construct diff")
+	}
+	log.Infof("Diff:\n---\n%s\n---\n", diffStr)
+}
+
+func readStreamMap(ocpBuildDataDir string, majorMinor majorMinor) (streamMap, error) {
+	streamMap := &streamMap{}
+	return *streamMap, readYAML(filepath.Join(ocpBuildDataDir, "streams.yml"), streamMap, majorMinor)
```

Comment:
Maybe this is just a personal preference, but I think I prefer to do the pointers like this:

```go
	streamMap := streamMap{}
	return streamMap, readYAML(filepath.Join(ocpBuildDataDir, "streams.yml"), &streamMap, majorMinor)
```

IMHO this is conceptually cleaner (and similar to the "borrowing" concept in Rust and how pointers are often thought about) -- we pass a reference to the function that needs to borrow the value and edit it, but we work with the value itself when we are creating and returning.